### PR TITLE
Added check port whether used by another application when starting HTTP Server

### PR DESCRIPTION
**Pull request by MuhammetDilmac**

### CODE REVIEW

**Comment by sdogruyol** [comment]
This looks like a good way to check before spawning the sever. Need to test it first :+1:

File: src/kemal/config.cr [hunk 1]
```diff
@@ -55,6 +58,16 @@ module Kemal
       HANDLERS.insert(0, @logger.not_nil!)
     end
 
+    def check_port
```
Comment (sdogruyol) [inline src/kemal/config.cr]: Needs to be `private`

File: src/kemal/config.cr [hunk 2]
```diff
@@ -55,6 +58,16 @@ module Kemal
...
+      begin
+        TCPSocket.new(@host_binding, @port).close
+        logger.write "#{@port} port is used by another process. Please select a different port.\n"
```
Comment (sdogruyol) [inline src/kemal/config.cr]: The message is better like this.

`Port 3000 is used by another process. Please specify another port with -p option when starting your Kemal application.`

**Comment by jmoriau** [comment]
`server.listen` already does exactly what you're doing.

```ruby
def listen
    server = TCPServer.new(@host, @port)
    until @wants_close
      spawn handle_client(server.accept)
    end
  end
```

Why not just wrap the `.listen` call in a begin/rescue and print the exception message, it would catch all the other potential errors.

**Comment by sdogruyol** [comment]
@jmoriau you are talking about `HTTP#Server` right? Monkey patching that doesn't feel right :(

**Comment by jmoriau** [comment]
Not monkey patching but this line in `kemal.cr` (33 I think):

```ruby
  config.logger.write "[#{config.env}] Kemal is ready to lead at #{config.scheme}://#{config.host_binding}:#{config.port}\n"
  server.listen
```

**Comment by sdogruyol** [comment]
That's within a `at_exit` block. I'm not sure if that exception gets rescued or swallowed there.

**Comment by jmoriau** [comment]
It  gets rescued.

**Comment by sdogruyol** [comment]
If it's so. We better handle this at `at_exit` hook like you mention @jmoriau

**Comment by jmoriau** [comment]
Would be better than creating a second `TCPServer` just to check if the port is available.

**Comment by sdogruyol** [comment]
Not `TCPServer`, a `TCPSocket` :)

**Comment by jmoriau** [comment]
Yeah my bad I typed a bit quickly.

**Comment by jmoriau** [comment]
Also since it's the last thing ever called, you could just leave it like this and if an exception happens, it will get printed to the console with a backtrace and stop the program.

**Comment by jmoriau** [comment]
Now that I think about this is also a half measure, it would be better to first work on removing the `at_exit` hook and have a better exception handling.

**Comment by sdogruyol** [comment]
@jmoriau both Sinatra and Kemal relies on `at_exit`.

**Comment by MuhammetDilmac** [comment]
In my opinion check port and bind options before at_exit block.

**Comment by jmoriau** [comment]
If you check before the at_exit there is no way to stop the program if something is wrong.

**Comment by sdogruyol** [comment]
Have you guys checked out the solution of Sinatra here? https://github.com/sinatra/sinatra/blob/4c7d38eb1b2cc02ce51029f28e0c3c34ca12ccfd/lib/sinatra/base.rb#L1080-L1094

**Comment by sdogruyol** [comment]
This is now fixed with https://github.com/sdogruyol/kemal/commit/fd904cd98d8dff92752b3484f6d7714477ddc6dd

**Pull request closed without merging**